_posts/2022-06-20-dask-multi-cloud.md

```diff
@@ -0,0 +1,1198 @@
+
+To do this, when we create the workers we mark them as having a `pool-ecmwf` or `pool-eumetsat` resource. Then when we want to create tasks that can only run in one data centre, we annotate them as requiring the appropriate resource
+
+
+```python
+with (dask.annotate(resources={'pool-ecmwf': 1})):
+    predictions.mean('realization').isel(height=10).compute()
+```
+
+A [special Python context manager](https://github.com/gjoseph92/dask-worker-pools) can hide the nastiness, allowing us to run a calculation only where the data is available
+
+
+```python
+with pool('ecmwf'):    
+    predictions.mean('realization').isel(height=10).compute()
+```
+
+Better still, we can load the data inside the context manager block and it will carry the task annonation with it
+
+
+```python
+with pool('ecmwf'):
+    predictions = xarray.open_dataset('/data/ecmwf/000490262cdd067721a34112963bcaa2b44860ab.nc').chunk('auto')
+```
+
+Another context manager in the library, `propagate_pools`, ensures that this resource pinning is respected
+
+
+```python
+with propagate_pools():
+    predictions.mean('realization').isel(height=10).compute()
+```
+
+This allows us to mark data with its pool
+
+
+```python
+with pool('ecmwf'):    
+    predictions = xarray.open_dataset('/data/ecmwf/000490262cdd067721a34112963bcaa2b44860ab.nc').chunk('auto')
+
+with pool('eumetsat'):
+    observations = xarray.open_dataset('/data/eumetsat/ad-hoc/observations.nc').chunk('auto')
+```
+
+define some deferred calculations oblivious to its provenance
+
+
+```python
+averaged_predictions = predictions.mean('realization')
+diff = averaged_predictions.isel(height=10) - observations
+```
+
+and then perform the final calculation
+
+
+```python
+%%time 
+with propagate_pools():
+    show_all_workers()
+    diff.compute()
+```
+
+    CPU times: user 127 ms, sys: 6.34 ms, total: 133 ms
+    Wall time: 4.88 s
+
+
+Remember, our aim was to distribute a calculation across data centres, whilst preventing workers reading foreign bulk data.
+
+Here, we know that data is only being read by workers in the appropriate location, because neither data centre can read the other's data. Once data is in memory, Dask prefers to schedule tasks on the workers that have it, so that the local workers will tend to perform follow-on calcuations.
+
+Ordinarily though, Dask would use idle workers to perform calculations even if they don't have the data. If allowed, this work stealing would result in unreduced data being moved between data centres, a potentially expensive operation, so the `propagate_pools` context manager also prevents work-stealing between workers in different pools.
```

Comment:
Ah I see some explaination about `propagate_pools` is here, maybe this would be better put when it is first mentioned.